auth/rest_http_client.hh

```diff
@@ -0,0 +1,197 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <seastar/net/api.hh>
+#include <seastar/net/dns.hh>
+#include <seastar/net/tls.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/future-util.hh>
+#include "picojson/picojson.h"
+#include "auth/rest_response_parser.hh"
+#include "auth/role_manager.hh"
+#include "alternator/base64.hh"
+
+
+namespace auth {
+
+    class rest_http_client {
+        std::string _server;
+        int _port;
+        std::string _ca_file;
+        shared_ptr <tls::certificate_credentials> _creds;
+        sstring _request;
+
+    public:
+        rest_http_client() {};
+
+        rest_http_client(std::string server, int port, std::string ca_file)
+                : _server(server), _port(port), _ca_file(ca_file),
+                  _creds(make_shared<tls::certificate_credentials>()) {}
+
+        class connection {
+        private:
+            connected_socket _fd;
+            input_stream<char> _read_buf;
+            output_stream<char> _write_buf;
+            rest_response_parser _parser;
+            sstring _request;
+
+            future <role_set> extract_groups(temporary_buffer<char> buf, uint content_len) {
+                const char *json = buf.get();
+                picojson::value v;
+                std::string err;
+                picojson::parse(v, json, json + content_len, &err);
+                if (!err.empty()) {
+                    std::throw_with_nested(exceptions::authentication_exception(
+                            format("Bad response: not able to parse body: {}", err)));
+                }
+
+                // check if the type of the value is "object"
+                if (!v.is<picojson::object>()) {
+                    std::throw_with_nested(exceptions::authentication_exception(
+                            "Bad response: response is not a dict"));
+                }
+
+                const picojson::value::object &obj = v.get<picojson::object>();
+                auto it = obj.find("groups");
+                if (it == obj.end()) {
+                    std::throw_with_nested(exceptions::authentication_exception(
+                            "Bad response: doesn't contain groups field"));
+                }
+
+                auto pv_groups = it->second;
+                if (!pv_groups.is<picojson::array>()) {
+                    std::throw_with_nested(exceptions::authentication_exception(
+                            "Bad response: groups field doesn't contain an array"));
+                }
+
+                const picojson::value::array &p_groups = pv_groups.get<picojson::array>();
+
+                role_set groups;
+                // TODO filter groups to add (For ex. only add groups containing scylla... to avoid storing all groups)
```

Comment:
I would implement group filtering to not have to come back on it later

Reply:
Why? Is it a real concern to have it in V0?

Reply:
not real concern